File: egg-niuke-training/18-permutations.py
"""
NC42 全排列
https://www.nowcoder.com/practice/4bcf3081067a4d028f95acee3ddcd2b1

给定一个不含重复数字的数组nums，返回其所有可能的全排列。

回溯算法
时间复杂度：O(n*n!)
空间复杂度：O(n)
"""

def permute(nums):
    """
    全排列 - 回溯法
    """
    result = []
    
    def backtrack(path, remaining):
        # 终止条件
        if not remaining:
            result.append(path[:])
            return
        
        # 选择列表
        for i in range(len(remaining)):
            # 做选择
            path.append(remaining[i])
            # 递归
            backtrack(path, remaining[:i] + remaining[i+1:])
            # 撤销选择
            path.pop()
    
    backtrack([], nums)
    return result

def permute_swap(nums):
    """
    全排列 - 交换法
    """
    result = []
    
    def backtrack(start):
        if start == len(nums):
            result.append(nums[:])
            return
        
        for i in range(start, len(nums)):
            # 交换
            nums[start], nums[i] = nums[i], nums[start]
            # 递归
            backtrack(start + 1)
            # 回溯
            nums[start], nums[i] = nums[i], nums[start]
    
    backtrack(0)
    return result

# 测试
if __name__ == "__main__":
    nums = [1, 2, 3]
    
    result = permute(nums)
    print(f"数组 {nums} 的全排列:")
    for perm in result:
        print(perm)
    
    print(f"\n共 {len(result)} 种排列")

File: egg-niuke-training/18-permutations.cpp
/**
 * NC42 全排列
 * https://www.nowcoder.com/practice/4bcf3081067a4d028f95acee3ddcd2b1
 * 
 * 给定一个不含重复数字的数组nums，返回其所有可能的全排列。
 * 
 * 时间复杂度：O(n*n!)
 * 空间复杂度：O(n)
 */

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    /**
     * 全排列 - 回溯法
     */
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> path;
        vector<bool> used(nums.size(), false);
        
        backtrack(nums, path, used, result);
        return result;
    }
    
    /**
     * 全排列 - 交换法
     */
    vector<vector<int>> permuteSwap(vector<int>& nums) {
        vector<vector<int>> result;
        backtrackSwap(nums, 0, result);
        return result;
    }

private:
    void backtrack(vector<int>& nums, vector<int>& path, 
                  vector<bool>& used, vector<vector<int>>& result) {
        // 终止条件
        if (path.size() == nums.size()) {
            result.push_back(path);
            return;
        }
        
        // 选择列表
        for (int i = 0; i < nums.size(); i++) {
            if (used[i]) continue;
            
            // 做选择
            path.push_back(nums[i]);
            used[i] = true;
            
            // 递归
            backtrack(nums, path, used, result);
            
            // 撤销选择
            path.pop_back();
            used[i] = false;
        }
    }
    
    void backtrackSwap(vector<int>& nums, int start, vector<vector<int>>& result) {
        if (start == nums.size()) {
            result.push_back(nums);
            return;
        }
        
        for (int i = start; i < nums.size(); i++) {
            swap(nums[start], nums[i]);
            backtrackSwap(nums, start + 1, result);
            swap(nums[start], nums[i]);
        }
    }
};

int main() {
    Solution solution;
    
    vector<int> nums = {1, 2, 3};
    vector<vector<int>> result = solution.permute(nums);
    
    cout << "数组 [1, 2, 3] 的全排列:" << endl;
    for (const auto& perm : result) {
        cout << "[";
        for (size_t i = 0; i < perm.size(); i++) {
            cout << perm[i];
            if (i < perm.size() - 1) cout << ", ";
        }
        cout << "]" << endl;
    }
    
    cout << "\n共 " << result.size() << " 种排列" << endl;
    
    return 0;
}

